src/lib.rs

```diff
@@ -438,338 +507,521 @@ impl TsunamiBuilder {
+                                                        );
+                                                        None
                                                     })
-                                                    .unwrap_or_default();
-                                                (instances, iid_to_name, false)
-                                            }),
-                                    )
+                                            })
+                                            .collect::<Vec<_>>()
+                                    })
+                                    .unwrap_or_default();
+                                WaitForInstancesResult::NotOk(instances)
+                            });
+                        Either::A(future)
+                    }
+                    Some(e) => Either::B(futures::future::err(e)),
+                }
+            });
+            Either::A(future)
+        } else {
+            Either::B(wait_for_instances)
+        }.and_then(|wait_result| {
+            if let WaitForInstancesResult::Ok(instances, id_to_name) = wait_result {
+                Either::A(futures::future::ok((id_to_name, instances)))
+            } else {
+                Either::B(futures::future::err(format_err!(
+                    "not all instances were launched within the time limit"
+                )))
+            }
+        })
+            .and_then(move |(id_to_name, instances)| {
+                // 3. wait until all instances are up
+                // note that we *don't* do this check if we have no instances, b/c empty instance list
+                // means "list all instances".
+                let mut desc_req = rusoto_ec2::DescribeInstancesRequest::default();
+                let term_instances = instances.clone();
+                desc_req.instance_ids = Some(instances);
+
+                futures::future::loop_fn(id_to_name, move |id_to_name| {
+                    ec2.describe_instances(&desc_req).map(move |reservations| {
+                        let mut machines = HashMap::new();
+                        for reservation in reservations.reservations.unwrap_or_else(Vec::new) {
+                            for instance in reservation.instances.unwrap_or_else(Vec::new) {
+                                match instance {
+                                    rusoto_ec2::Instance {
+                                        instance_id: Some(instance_id),
+                                        instance_type: Some(instance_type),
+                                        private_ip_address: Some(private_ip),
+                                        public_dns_name: Some(public_dns),
+                                        public_ip_address: Some(public_ip),
+                                        ..
+                                    } => {
+                                        let name = id_to_name[&instance_id].clone();
```

Comment:
I don't think it should be necessary to `clone` here? Just stick to `&*id_to_name[&instance_id]`.